pkg/operator/staticpod/startupmonitor/startup_monitor.go

```diff
@@ -0,0 +1,370 @@
+	defer klog.Info("Shutting down the startup monitor")
+
+	if err := m.healthChecker.Start(ctx); err != nil {
+		return fmt.Errorf("failed to start the health checker for %s, due to %v", m.targetName, err)
+	}
+
+	wait.Until(m.syncErrorWrapper, m.probeInterval, ctx.Done())
+	return nil
+}
+
+func (m *monitor) syncErrorWrapper() {
+	if err := m.sync(); err != nil {
+		klog.Error(err)
+	}
+}
+
+func (m *monitor) sync() error {
+	//
+	// TODO: acquire an exclusive lock to coordinate work with the installer pod
+	//
+	// a lock is required to protect the following case:
+	//
+	// an installer is in progress and wants to install a new revision
+	// the current revision is not healthy and we are about to fall back to the previous version (fallbackToPreviousRevision method)
+	// the installer writes the new file and we immediately overwrite it
+	//
+	// additional benefit is that we read consistent operand's manifest
+
+	// to avoid issues on startup and downgrade (before the startup monitor was introduced check the current target's revision.
+	// refrain from any further processing in case we have a mismatch.
+	currentTargetRevision, err := m.loadRootTargetPodAndExtractRevision()
+	if err != nil {
+		return err
+	}
+	if m.revision != currentTargetRevision {
+		klog.Infof("Stopping further processing because the monitor is watching revision %d and the current target's revision is %d", m.revision, currentTargetRevision)
+		return nil
+	}
+
+	if m.monitorTimeStamp.IsZero() {
+		m.monitorTimeStamp = time.Now()
+	}
+
+	// first check if the target is healthy
+	// note that we will always reconcile on transient errors
+	// before starting the fall back procedure
+	healthy, reason, err := m.healthChecker.IsTargetHealthy()
+	if healthy {
+		klog.Info("Observed a healthy target, creating last known good revision")
+		if err := m.createLastKnowGoodRevisionAndDestroy(); err != nil {
+			return err
+		}
+		return nil
+	}
+	if err != nil {
+		klog.Infof("failed to assess health condition due to err = %v, reason = %v", err, reason)
+	}
+
+	// check if we reached the timeout
+	if time.Now().After(m.monitorTimeStamp.Add(m.timeout)) {
+		klog.Infof("Timed out while waiting for the target to become healthy, starting a fallback procedure. Timeout is %v, MonitorTS is %v, CurrentTS is %v", m.timeout, m.monitorTimeStamp, time.Now())
+		// TODO: report reason and err
+		_ = reason
+		_ = err
+		if err := m.fallbackToPreviousRevision(); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	return nil
+}
+
+func (m *monitor) createLastKnowGoodRevisionAndDestroy() error {
+	// step 0: rm the previous last good known revision if exists
+	// step 1: create last known good revision
+	if err := m.createLastKnowGoodRevisionFor(m.revision, true); err != nil {
+		return err
+	}
+
+	// step 2: commit suicide
+	return m.io.Remove(path.Join(m.manifestsPath, fmt.Sprintf("%s-startup-monitor.yaml", m.targetName)))
```

Comment:
would move all this logic independent of the readiness monitor out of this file, and call it in the outer Run func of the command.

Reply:
Also the suicide should happen on both case: timeout and ready operand.

Reply:
done